hibernate-core/src/main/java/org/hibernate/persister/entity/SingleTableEntityPersister.java

```diff
@@ -349,10 +351,10 @@ public SingleTableEntityPersister(
 			);
 
 			// SUBCLASSES
-			final List<Subclass> subclasses = persistentClass.getSubclasses();
+			final List<Subclass> subclasses = persistentClass.getSubclasses().stream().sorted(Comparator.comparing(PersistentClass::getEntityName)).collect(Collectors.toList());
 			for ( int k = 0; k < subclasses.size(); k++ ) {
 				Subclass subclass = subclasses.get( k );
-				subclassClosure[k] = subclass.getEntityName();
+				subclassClosure[k+1] = subclass.getEntityName();
```

Comment:
This off by one error was accidentally introduced in commit [a7da407](https://github.com/hibernate/hibernate-orm/commit/a7da40709ca30ad5000fc7daa5f47b7b69bdc735#diff-40b4cbf60f5729945ece457b8d8e2ee5ad97bdf2606868fdfc537f24aa07f7fcR404) during a refactor from a foreach loop to a for loop.